tests/unit/lms/services/canvas_test.py

```diff
@@ -1,30 +1,370 @@
-from unittest.mock import sentinel
+from unittest.mock import DEFAULT, call, sentinel
 
 import pytest
 
-from lms.services import CanvasFileNotFoundInCourse, CanvasService
+from lms.services import (
+    CanvasAPIPermissionError,
+    CanvasFileNotFoundInCourse,
+    CanvasService,
+)
 from lms.services.canvas import factory
 from tests import factories
```

Comment:
I think these tests really need some thought into how they can be more concise and easier to read. Many of the tests are longer than my monitor vertically. There's a lot of duplication here and it's not clear to me that we're testing different scenarios, or what is different between scenarios.

I think we could do with unit tests which have a narrower non-overlapping focus and attempt to assert one or two things about the code. I feel like we've ended up with a Cartesian product of all possible combinations of events as scenarios rather than unit tests isolating those different events.

I don't think we need to repeat and assert every call the code makes in each scenario. We should focus in on what's different about this one. For example, with the `check_in_course` tests we should prove we call those functions relating to `check_in_course`. We don't need to re-assert all of the regular calls and return values we've covered in other tests.

In general I would also favour anything that would make this shorter. Some of the fixtures like `file_id_from_current_course` would probably be better being replaced with `file_from_current_course["id"]`. `canvas_api_client` could be replaced with `canvas_service.api` which mirrors the usage in the code. A `functools.partial` around `public_url_for_file` would cut down on the number of fixtures in every test, and allow you to indicate what's important about this scenario by only specifying the important parts of the call that are relevant in this scenario. 

`module_item_configuration` is probably better referred to as `assignment` throughout. It is the return value of the `assignment` service and is more concise and easier to read.

I suspect the level of comment here is an attempt to make this easier to understand, but I'd suggest it's a sign that the tests are too confusing in the first place. Either this level of comment isn't really necessary and should be reduced for readability, or the tests are probably too complicated, and should be simplified so the comments aren't required.

Reply:
I think it might be worth going through the exercise of mapping out the planned tests as a skeleton test file (without any content in the test) just to get the names and design down. Something like...

```python
class TestPublicURLForFile:
    def test_it(self):
        # Assert call the Canvas API with the id
        # Assert we return the value
        ...

    def test_it_uses_a_file_mapping_if_present(self):
        # Assert we call the assignment to get a mapping
        # Assert call the Canvas API with the mapped id
        ...

    def test_check_in_course_looks_up_the_file_in_the_course(self):
        # Test we list the files of the correct course
        ...

    def test_check_in_course_raises_if_the_file_is_not_found(self):
        # Assert we raise
        ...

    @pytest.mark.usefixtures("with_failing_lookup")
    def test_it_can_remap_a_file(self):
        # Setup a file record
        # Setup a mapping with a matching file

        # Assert we store the new mapping
        ...

    @pytest.mark.usefixtures("with_failing_lookup")
    def test_it_cannot_map_a_file_without_having_seen_it_before(self):
        # Assert we blow up
        ...

    @pytest.mark.usefixtures("with_failing_lookup")
    @pytest.mark.parametrize("field,value", (("name", "different"), ("size", "99999")))
    def test_it_cannot_map_a_file_if_there_is_no_matching_file(self, field, value):
        # Setup a mapping with a matching file

        # Assert we blow up
        ...

    @pytest.fixture(params=(CanvasFileNotFoundInCourse, CanvasAPIPermissionError))
    def with_failing_lookup(self, param):
        # Configure the situation so we raise the correct error
        ...
```

Reply:
I've had a go at reworking these tests: changing test names, removing comments, getting rid of unnecessary tests.

I'm not sure there's much duplication left in these that's worth removing. There is code that superficially looks like duplication because some of these tests do similar-but-different assertions. I think trying to deduplicate that might make things worse rather than better.

This is what I think needs to be covered:

* `check_in_course` can either be `True` or `False`. There needs to be a happy-path test for each case:
  - [x] `check_in_course=True`: `test_if_check_in_course_is_True_it_checks_that_the_file_is_in_the_course()`
  - [x] `check_in_course=False`: `test_without_check_in_course_and_without_a_mapped_file_id()` 
* There can either be a `mapped_file_id` in the DB or not. There needs to be a happy-path test for each case
  - [x] With a mapping in the DB: `test_if_theres_a_mapped_file_id_it_uses_it()`
  - [x] Without a mapping in the DB: This was already covered by earlier tests like `test_without_check_in_course_and_without_a_mapped_file_id()` 
* I don't think we need to cross-product the `check_in_course` happy paths with the `mapped_file_id` ones but we do need tests for when `check_in_course` is `True` and there's no `mapped_file_id` (it should check for the original `file_id` in the course) and for when `check_in_course` is `True` and there _is_ a `mapped_file_id` (it should check for the `mapped_file_id` in the course, not the original `file_id`)
  - [x] `check_in_course=True` and there's no mapping: covered by the previously mentioned `test_if_check_in_course_is_True_it_checks_that_the_file_is_in_the_course()` 
  - [x] `check_in_course=True` and there _is_ a mapping: covered by the previously mentioned `test_if_theres_a_mapped_file_id_it_uses_it()`
* If `check_in_course` is `True` then `assert_file_in_course()` can raise `CanvasFileNotFoundInCourse`. Also `public_url()` can raise `CanvasAPIPermissionsError`, realistically I think this only happens when `check_in_course` is `False`. We need to test that when _either_ `CanvasFileNotFoundInCourse` or `CanvasAPIPermissionsError` is raised:
  - [x] If there's no `File` object in the DB it raises: `test_file_not_found_in_course_but_no_file_info()` and `test_permissions_error_but_no_file_info()`
  - [x] If there is a file in the DB but there's no matching file in the Canvas course, it raises: `test_file_not_found_in_course_but_no_matching_file()` and `test_permissions_error_but_no_matching_file()`
  - [x] If there's both a file in the DB and a matching file in the Canvas course then it saves a mapping and returns the public URL of the matching file: `test_file_not_found_in_course_and_matching_file_found()` and `test_permissions_error_and_matching_file_found()`

All of the tests that're now remaining in the file are needed to cover those above cases. These cases don't literally cover everything that could possibly happen. But I think they give good enough coverage, and cover the most common cases.

* * *

> Some of the fixtures like `file_id_from_current_course` would probably be better being replaced with `file_from_current_course["id"]`

It's actually `str(file_from_current_course["id"])` and `str(file_from_another_course["id"])` every time, which can be a bit of a mouthful, but I've gone ahead and made this change.

> `canvas_api_client` could be replaced with `canvas_service.api`

The downside of this is that it couples the tests more closely to the `CanvasService` code. If `CanvasService.api` changed lots of tests would break. I've gone ahead and made this change anyway.

> A `functools.partial` around `public_url_for_file` would cut down on the number of fixtures in every test, and allow you to indicate what's important about this scenario by only specifying the important parts of the call

This actually doesn't cut down on fixtures much: in most tests it just replaces the `module_item_configuration` fixture with a `public_url_for_file` fixture. In some tests it actually increases the number of fixtures as `module_item_configuration` is still needed anyway. But it does cut down on the arguments that need to be passed to `public_url_for_file()`, allowing each test to pass only the relevant arguments. I've gone ahead and made this change.

* * *

The longest of these tests is now 33 lines, but it's actually only 19 lines of code (including blank lines) if you don't count comments and the argument list. The tests are all quite simple in form too.

* * *

I've cut down the comments somewhat. But I think the comments here are appropriate. The comments are not explaining how the _tests_ work. They're not there because the tests are confusing. They're explaining the complex external scenarios that are being covered. Example:

```python
    def test_file_not_found_in_course_but_no_matching_file(
        self, file_service, file_from_a_different_course, public_url_for_file
    ):
        # This is what happens when a teacher launches an assignment whose
        # file_id is *not* in the current course and even though we do have a
        # record of this file_id in our DB we don't find a matching file in the
        # current course. Since we can't find a matching file we can't fix the
        # assignment so we raise an error.

        # The file record that we find in our DB. Its name doesn't match any
        # file in the current course.
        file_service.get.return_value = factories.File(name="foo")

        with pytest.raises(CanvasFileNotFoundInCourse):
            public_url_for_file(
                file_id=str(file_from_a_different_course["id"]), check_in_course=True
            )
```